Merge mircea's latest uncommited script calls script

MirciuPirciu opened the pull request:
I modified the script:
* New dependency (shutil), for deleting generated Ghidra Projects (as i didn't find a flag or anything easier to flag the pyghidra to make them temporary). If there's an easier way lmk.

* Added a function to generate the ICFG as an nx.Digraph() Object. It also returns the entry address of the executable.
** The ICFG gen. function also translates external addresses to external function names for CALL instructions and nodes.
** ICFG gen. function dumps a dot file for visualization purposes. May be omitted

* For parse_sequence(), the function was modified. The sequences start from the entry node (at the entry address). We no more generate trees for every node with internal grade 0.

* I/O wise: the script looks for executables inside a directory named 'binaries'. For each it constructs the ICFG, dumps it in a DOT File, and the sequences are outputted in a main folder called 'sequences' with sub directories for each  binary.

raduicaiulian committed:
removed leftovers + call script since there are so little of it that we need it make sense to start from scratch

raduicaiulian committed:
addeed Makefile, Mircea's script+sample, immproved the script a little bit

MirciuPirciu committed:
Rectify Script

Modified the main script with more automatic features for the methodology. This includes:

* New dependency (shutil), for deleting generated Ghidra Projects (as i didn't find a flag or anything easier to flag the pyghidra to make them temporary)

* added a function to generate the ICFG in an nx.Digraph() Object. It also returns the entry address of the executable.

** The ICFG gen. function also translates external addresses to external function names  for CALL instructions and nodes.

** ICFG gen. function dumps a dot file for visualization purposes. May be omitted

* For parse_sequence(), the function was modified. The sequences start from the entry node (at the entry address). We no more generate trees for every node with internal grade 0.

* I/O wise: the script looks for executables inside a binary named 'binaries'. For each it constructs the ICFG, dumps it in a DOT File, and the sequences are outputted in a main folder called 'sequences' with sub directories for each  binary.

Updated a bit the requirements.txt and README.md

raduicaiulian commented:
Regrading the fourth point, I really don't like the fact that we have hardcoded paths in the program, we should read them from environment variables;

raduicaiulian commented on main.py:
```diff
@@ -1,53 +1,210 @@
+
+	#Go to all the CFGs and generate all the sequences
+	#TODO Optimize. Write sequence in file once it is done.
+	fls = os.scandir('./binaries') #Scan directory
```
We shouldn't scan the directory, it is not essential for this script, if running this script for multiple binaries is desired then the script should be run multiple times, with each binary passed via parameter as an environment variable.

raduicaiulian replied:
You should have used the environment variables form that are defined in Makefile(eventually update them with the actual values for you environment).
I got the feeling that you didn't actually understood my previous commit .

raduicaiulian replied:
You shouldn't have removed:
`with pyghidra.open_program(os.environ['PATH_TO_PROGRAM_TO_ANALYSE']) as flat_api:`

raduicaiulian commented on main.py:
```diff
@@ -1,53 +1,210 @@
+	for f in fls: # For each DOT file
+		os.system('mkdir ./sequences/{}'.format(f.name))	
+		logger.info(">>> Executing on {}..".format(f.name))
+		parse_sequences(f.name) # Do the thing..
```
This comment is useless, you should either remove or update it, I think removing it is fine.

MirciuPirciu commented:
Wouldn't it be better to have them as arguments instead?

raduicaiulian commented on main.py:
```diff
@@ -1,53 +1,210 @@
+	from ghidra.app.util.bin.format.pe import OptionalHeader
+	from ghidra.app.util.bin.format.pe import PortableExecutable
+	#tardir = sys.argv[1]
+	os.system('mkdir ./sequences')
```
We can ensure using git that this directory is already there.

raduicaiulian replied:
It turns out I was wrong, we agreed to keep the old solution until we find something better.

raduicaiulian commented on main.py:
```diff
@@ -1,53 +1,210 @@
+		os.system('mkdir ./sequences/{}'.format(f.name))	
+		logger.info(">>> Executing on {}..".format(f.name))
+		parse_sequences(f.name) # Do the thing..
+		shutil.rmtree("./binaries/{}_ghidra".format(f.name))
```
You should replace this line with **shutil.rmtree(f"./binaries/{f.name}_ghidra")**, it is more concise.
Do that for all places where you used **format** function

raduicaiulian commented on requirements.txt:
```diff
@@ -2,3 +2,4 @@ networkx==3.4.2
 pyghidra==2.0.1
 pandas
 pydot
```
You should check on you local setup what are the actual versions(pip show <lib> | grep "Version:") you are using right now and set them here, because using the latest version is a sure way to break the code sometime in the future.

MirciuPirciu replied:
I added the version in requirements.txt.
shutil doesn't have a version it seems.

raduicaiulian commented on main.py:
```diff
@@ -1,53 +1,210 @@
+	from ghidra.app.util.bin import MemoryByteProvider
+	from ghidra.app.util.bin.format.pe import OptionalHeader
+	from ghidra.app.util.bin.format.pe import PortableExecutable
+	#tardir = sys.argv[1]
```
Remove it.

raduicaiulian commented on main.py:
```diff
@@ -1,53 +1,210 @@
+def gen_icfg(fd):
+	icfg = nx.DiGraph() 
+
+	with pyghidra.open_program("./binaries/{}".format(fd)) as flat_api: #Analyze the binary file
```
See https://github.com/MirciuPirciu/CFG_Malware_Script_Thingy/pull/8/commits/85bae00b5f6ae1f0f0b7e518d63f660ec309fbcd#r1997247546

raduicaiulian committed:
used environment variable + upgraded the Makefile and Readme

MirciuPirciu commented on main.py:
```diff
@@ -0,0 +1,209 @@
+from networkx.drawing.nx_pydot import write_dot 
+
+logger = logging.getLogger(__name__) # TO DO: this should be improved in the future to keep the function name
+logging.basicConfig(filename='example.log', encoding='utf-8', level=logging.DEBUG)
```
Change 'example.log' with, for instance,  'example_<binary_name>.log'. Might be useful to separate logs for each binary.

raduicaiulian commented on main.py:
```diff
@@ -0,0 +1,209 @@
+
+		icfg = nx.relabel_nodes(icfg, relabel_mapping) #Relabel
+
+		#Get the Entry Point of the Portable Executable. Save it inside a global var.
+		logger.info("Getting Entry Address")
+		byteProvider = MemoryByteProvider(program.getMemory(),program.getImageBase())
+		pe = PortableExecutable(byteProvider, PortableExecutable.SectionLayout.MEMORY)
+		optHeader = pe.getNTHeader().getOptionalHeader()
+		entry_address = hex(optHeader.getAddressOfEntryPoint() + program.getImageBase().getOffset()).replace('x','0')
+		entry_address = '"'+entry_address+'"'
+		logger.info("Entry at {}".format(entry_address))
```
This should be encapsulated in a function called **getEntryPoint**.

MirciuPirciu committed:
Rectified some comments

MirciuPirciu committed:
function and few modificaitons

1. Made the entry function computation as a function

2. Fixed formatting for the entry address.

3. Added shutil.rmtree for removing projects until we find an elegant way to make them temporary through pyghidra

MirciuPirciu commented on Makefile:
```diff
@@ -1,9 +1,9 @@
 run: export GHIDRA_INSTALL_DIR=/home/i/Downloads/ghidra_11.3_PUBLIC
 run: export PATH_TO_PROGRAM_TO_ANALYSE=/home/i/repo/CFG_Malware_Script_Thingy/artifacts/samples/elf_example_binary.bin
-run: export OUTPUT_DOT_FILE=./artifacts/output/icfg.dot
+run: export OUTPUT_DOT_FILE=./artifacts/output/{}_icfg.dot
```
I suggest we use a single environment variable for a general output folder where we put the DOT file and the sequence files. 

So when the script runs, inside the output directory we would have:
output_directory:
* program_icfg.dot
* program_sequences
** program_sequence_0.txt
** program_sequence_1.txt 
...

raduicaiulian commented on main.py:
```diff
@@ -104,23 +117,17 @@ def gen_icfg(fd):
 
 		logger.info("ICFG Built..")
-		write_dot(icfg, os.environ['OUTPUT_DOT_FILE'].format(program.getName())) #Dump the ICFG into a dot
+		write_dot(icfg, os.environ['OUTPUT_DOT_FILE'] + '/{}'.format(program.getName())) #Dump the ICFG into a Graphviz DOT File
```
Why did you added **+ '/{}'**, because in the `Makefile` **OUTPUT_DOT_FILE** is defined as `OUTPUT_DOT_FILE=./artifacts/output/{}_icfg.dot`?
Did you tested these changes?

MirciuPirciu replied:
Oops, my bad. I will rectify it.

MirciuPirciu committed:
Resolving Reviews

* Using f"" instead of format
* Fixed path generations
* Modified reuirements.txt
* Modified to use env. variables for paths

MirciuPirciu committed:
Fixes and few modifications

* Fixed path generation. Path are now generated correctly from the Tree generated from the ICFG
* Modified file output locations
* Tree DOT file is now dumped too for visualization
* Log is now named after the program name
* Changed the name of the sequence files
* Fixed entry point address retrieval. Now we are padding with zeros any addressed less than 8 bytes

raduicaiulian committed:
fixed problem with shutils, it turns out that you were unable to find a version for this package because there was an 's' missing in the name of the package
